robocorp-code/src/robocorp_code/inspector/java/robocorp_java/_inspector.py

```diff
@@ -0,0 +1,235 @@
+        "tree": ContextNode,
+    },
+)
+
+TFun = TypeVar("TFun", bound=Callable[..., Any])
+
+
+class ElementInspector:
+    def __init__(self) -> None:
+        from robocorp_code.inspector.windows.robocorp_windows import WindowElement
+        from robocorp_code.inspector.java.picker import CursorListenerThread
+
+        # TODO: the context now stores only the latest searched ContextNode and it's children
+        # to the given search depth. When user traverses back up the tree we would need to have
+        # the cached snapshot of all the previous searches in store.
+        #
+        # We need to store the whole tree here and always just update the tree from the locator
+        # the user is picking and update the matched node and it's children.
+        #
+        # This will also need an update to the JavaAccessBridgeWrapper to have an API to update only
+        # it and it's children (refresh is there, but does it respect search depth?).
+        from ._event_pump import EventPumpThread
+
+        self._context: Optional[ContextNode] = None
+        self._selected_window: Optional[str] = None
+        self._event_pump_thread: Optional[EventPumpThread] = None
+        self._jab_wrapper: Optional[JavaAccessBridgeWrapper] = None
+        self._window_obj: Optional[WindowElement] = None
+        self._picker_thread: Optional[CursorListenerThread] = None
+
+    @property
+    def event_pump_thread(self):
+        if not self._event_pump_thread:
+            from ._event_pump import EventPumpThread
+
+            self._event_pump_thread = EventPumpThread()
+            self._event_pump_thread.start()
+
+        return self._event_pump_thread
+
+    @property
+    def jab_wrapper(self):
+        try:
+            if not self._jab_wrapper:
+                self._jab_wrapper = self.event_pump_thread.get_wrapper()
+            return self._jab_wrapper
+        except Exception as e:
+            log.error(e)
+            self.event_pump_thread.stop()
+            self._event_pump_thread = None
+            self._jab_wrapper = None
+
+    def list_windows(self) -> List[JavaWindow]:
+        return self.jab_wrapper.get_windows()
+
+    def set_window(self, window: str) -> None:
+        from robocorp_code.inspector.windows.robocorp_windows import desktop
+
+        self._selected_window = window
+        self.jab_wrapper.switch_window_by_title(self._selected_window)
+        self._window_obj = desktop().find_window(
+            f"{window}",
+            search_depth=1,
+            foreground=True,
+            move_cursor_to_center=False,
+        )
+
+    def bring_app_to_frontend(self):
+        import time
+
+        self._window_obj.update_geometry()
+        if not self._window_obj.has_valid_geometry():
+            self._window_obj.restore_window()
+            # compensating because of Windows animations
+            time.sleep(0.25)
+        self._window_obj.foreground_window(move_cursor_to_center=False)
+        time.sleep(0.2)
```

Comment:
The first sleep seems fine, but why the 2nd one is needed? i.e.: Deserves a comment on what fails if it's removed...

Reply:
Added comment